---
title: "Multi-Population Dose Response Study"
author: "Jane Park"
date: "2/6/2019"
output: 
  html_notebook: 
    fig_caption: yes
    toc: yes
  html_document: 
    keep_md: yes
    toc: yes
editor_options: 
  chunk_output_type: console
---

2019 Feb 6

Jane Park

## Libraries

```{r libraries}
library(readxl)
library(tximportData)
library(tximport)
library(lattice)
library(biomaRt)
library(readr)
library(gplots)
library(RColorBrewer)
library(edgeR)
library(limma)
library(tidyr)
library(dplyr)
library(data.table)

source("~/Documents/NIEHS_LSU_UCD/scripts/fun_MeanVar_all.R")
source("~/Documents/NIEHS_LSU_UCD/scripts/fun_TreatMeans.R")

setwd("~/Documents/NIEHS_LSU_UCD/rnaseq/")

```


## Import quantification tables 

Name quantification files path: 
```{r}
dir <- "~/Documents/NIEHS_LSU_UCD/rnaseq/190206_multipop"
files_list <- list.files(dir)
files <- file.path(dir,files_list, "quant.sf")
all(file.exists(files))
names <- gsub("_quant|_USP.*", "", files_list)
names(files) <- names
```

## Associate transcripts to gene IDs 

I'm using the gene model annotation file from supplementary materials of Noah Reid's 2016 paper, ["The genomic landscape of rapid repeated evolutionary adaptation to toxic pollution in wild fish"](http://science.sciencemag.org/content/suppl/2016/12/07/354.6317.1305.DC1)

Subset the appropriate columns from the above table. 

"We first make a data.frame called tx2gene with two columns: 1) transcript ID and 2) gene ID. The column names do not matter but this column order must be used. The transcript ID must be the same one used in the abundance files.""
```{r}
gene_names <- read.csv("~/Documents/NIEHS_LSU_UCD/rnaseq/Fhet_gene_transcript_names.csv")
cols <- c("row", "transcript_id", "gene_id")
colnames(gene_names) <- cols
tx2gene <- gene_names[,2:3]
head(tx2gene)
```


## Convert transcript-level counts to gene-level counts, and save as a DGEList object for easy handling. Counts are scaled for library size.
txi$abundance is the TPM table for gene-level estimates. txi$counts is the counts normalized for transcript lengths. 

```{r}
txi <- tximport(files, type="salmon", tx2gene=tx2gene, countsFromAbundance = "lengthScaledTPM")
names(txi)
head(txi$counts)

y <- DGEList(txi$counts)
head(y$samples)
```

```{r Extract sample info from names}
samplenames <- rownames(y$samples)
y$samples[,"exposure"] <- NA
y$samples[,"pop"] <- NA
y$samples[,"trt"] <- NA

for (i in 1:nrow(y$samples)) {
  if (nchar(samplenames)[i] == 13) {
    y$samples$exposure[i] <- substring(samplenames[i], 9,10)
    y$samples$pop[i] <- substring(samplenames[i], 5, 7)
    
  } else if (nchar(samplenames)[i] == 12) {
    y$samples$exposure[i] <- substring(samplenames[i], 6,7)
    y$samples$pop[i] <- substring(samplenames[i], 1, 3)
    
    } else if (nchar(samplenames)[i] == 11) {
      y$samples$exposure[i] <- substring(samplenames[i],5,6)
      y$samples$pop[i] <- substring(samplenames[i], 1, 2)
      
    }
}

y$samples$trt <- with(y$samples, paste(pop, exposure, sep="."))

y$samples$exposure <- as.factor(y$samples$exposure)
y$samples$pop <- as.factor(y$samples$pop)
y$samples$trt <- as.factor(y$samples$trt)
 
```

## Filtration

```{r}
#Save unfiltered, untransformed data as another object, save as .csv file:
zz <- y	
write.table(zz$counts, file="~/Documents/NIEHS_LSU_UCD/niehs/multipop/190206_unfiltered_counts.csv",
	col.names=TRUE, row.names=TRUE, sep="\t")
write.table(zz$samples, file="~/Documents/NIEHS_LSU_UCD/niehs/multipop/190206_all_samples.csv",
	col.names=TRUE, row.names=TRUE, sep="\t")
```


### Filter out lowly expressed reads

Filter all samples based on counts greater than 5 in at least 4 replicate samples of 1 treatment group: 

```{r}
#Create a list of genes for which there are greater than 5 counts in at least 4 replicates for each treatment group: 
d = list()
treat <- unique(y$samples$trt)
head(treat)

for (i in 1:length(treat)) {
	a <- which(y$samples$trt==treat[i])
	b <- colnames(y$counts[,a])
	kp <- rowSums(y$counts[,a] > 5) >=4 
	g <- which(kp==TRUE)
	d[[i]] <- g
}

#Subset the DGEList object y by the gene list created above:
nrow(y$counts)
filt_genes <- unique(names(unlist(d, use.names=TRUE)))
length(filt_genes)
y <- y[filt_genes,]
nrow(y$counts)

#Save filtered background gene list:
write.table(filt_genes, file="~/Documents/NIEHS_LSU_UCD/niehs/multipop/190206_background_gene_list.csv",
	col.names=FALSE, row.names=FALSE, sep="\t")
```

## Normalize and Log Transform Read Counts

Optimal constant for rooted log transformation is around 15.

```{r}
trans <- log2(y$counts+15)
```

#### Normalize log-transformed counts for all samples: 

```{r}
mean_counts <- colMeans(trans)
norm_counts <- sweep(trans, 2, colMeans(trans), "-")
grandmean <- mean(mean_counts)
norm_counts <- norm_counts + grandmean

```

Calculate mean and standard deviation of each gene for every treatment after normalization, whole data set: 

```{r}
ttrans <- t(norm_counts)
ttrans <- as.data.frame(ttrans)
ttrans$trt <- y$samples$trt

mv <- MeanVar(ttrans)
mea <- as.vector(mv$mean)
sd <- as.vector(mv$stdev)
slope <- (coef(lm(sd~mea))[2])
print(slope)

```

Visualize mean-standard deviation relationship after normalization. 
```{r}
pdf('~/Documents/NIEHS_LSU_UCD/niehs/multipop/190206_normcountslog+15.pdf')
plot(mea,sd, pch=16, cex=0.5, main="Mean-SD relationship of Normalized Log(counts + 15), slope=-.027")
abline(coef(lm(sd~mea)),col="red",lwd=3)
lines(lowess(mea,sd,f=0.2),col="blue",lwd=3)
dev.off()

```

If everything looks ok, save normalized counts as DGEList object y. 
```{r}
y$counts <- norm_counts
```


## Unsupervised clustering of Data/ MDS plot:

```{r}
# MDS Plot
par(mfrow=c(1,1))
treats <- y$samples$trt
exp <- y$samples$exposure
pop <- y$samples$pop

col.exp <- as.factor(exp)
levels(col.exp) <- brewer.pal(nlevels(col.exp), "Set1")
col.exp <- as.character(col.exp)

col.treat <- as.factor(treats)
levels(col.treat) <- brewer.pal(nlevels(col.treat), "Set1")
col.treat <- as.character(col.treat)

col.pop <- as.factor(pop)
levels(col.pop) <- brewer.pal(nlevels(col.pop), "Set2")
col.pop <- as.character(col.pop)

plotMDS(y, labels=rownames(y$samples), top=10000, col=col.pop, dim=c(1,2))

w <- y[,rownames(y$samples)!="0020ARSC10351" & rownames(y$samples)!="0093ARSC00353"]
plotMDS(y, labels=y$samples$trt, top=24977, col=col.pop, dim=c(1,2))
```

## Save all log-transformed & normalized counts and sample information for whole data set and subsets: 

```{r}
#Normalized counts and samples for whole data set: 
write.table(y$counts, "~/Documents/NIEHS_LSU_UCD/niehs/multipop/190206_allstages_no_outlier_normcounts_noouts.csv", 
		col.names=TRUE, row.names=TRUE, sep=",")
write.table(y$samples, "~/Documents/NIEHS_LSU_UCD/niehs/multipop/190206_allstages_no_outlier_samplesinfo_noouts.csv", 
    col.names=TRUE, row.names=TRUE, sep=",")

```

## DEG Analysis

```{r}
#Set up data for lm(): 
ttrans$exposure <- y$samples$exposure
ttrans$pop <- y$samples$pop

#Set up data frame to store p-values: 
d <- dim(ttrans)[2]-3
glist.y <- colnames(ttrans[,1:d])
cofs <- c("pop", "exposure", "pop:exposure")
pdat.y <- matrix(nrow=length(glist.y), ncol=length(cofs))
colnames(pdat.y) <- cofs
rownames(pdat.y) <- glist.y

#Run lm() on each gene: 
for (i in 1:ncol(ttrans[,1:d])) {
	g <- which(rownames(pdat.y)==colnames(ttrans)[i])
	l <- lm(formula=ttrans[,i] ~ pop*exposure, data=ttrans)
	a <- anova(l)
	p <- a$`Pr(>F)`[1:3]
	pdat.y[g,] <- p 
}

#Save p-values: 
pdat.y <- as.data.frame(pdat.y)
names(pdat.y) <- gsub(":", "_", names(pdat.y))
write.table(pdat.y, "~/Documents/NIEHS_LSU_UCD/niehs/multipop/190207_multipop_anova_unadj_pvalues.csv", 
		col.names=TRUE, row.names=TRUE, sep=",")

```

#### Adjust p-values for FDR

```{r}
#Vectorize and apply p.adjust to pdat.y dataframe
pval_all <- read.csv("~/Documents/NIEHS_LSU_UCD/niehs/multipop/190207_multipop_anova_unadj_pvalues.csv", header=TRUE)

adj_pval_all <- setDT(pval_all, keep.rownames=TRUE)
colnames(adj_pval_all)[1] <- "Gene"
adj_pval_all <- adj_pval_all %>% 
	gather(`pop`, `exposure`, `pop_exposure`,
		key = "cof", value = "pvals")

adj_pval_all$adj_pvals <- p.adjust(adj_pval_all$pvals, method="fdr")

#Filter pvals by DEG (p < 0.05): 
deg <- adj_pval_all[adj_pval_all$adj_pvals <= 0.05,]

#Write adjusted pvalues to csv:
write.table(adj_pval_all, "~/Documents/NIEHS_LSU_UCD/niehs/multipop/190207_fdradj_pval.csv",
		col.names=TRUE, row.names=FALSE, sep=",")

write.table(deg, "~/Documents/NIEHS_LSU_UCD/niehs/multipop/190207_fdradj_p005_DEG.csv",
		col.names=TRUE, row.names=FALSE, sep=",")
```

### Save DEG unique to main or interaction effects for each stage: 

```{r}
a <- deg[deg$cof == "pop",]
b <- deg[deg$cof == "exposure",]
c <- deg[deg$cof == "pop_exposure",]

a2 <- a[!(a$Gene %in% b$Gene) & !(a$Gene %in% c$Gene) ,]
b2 <- b[!(b$Gene %in% c$Gene) & !(b$Gene %in% a$Gene) ,]
c2 <- c[!(c$Gene %in% a$Gene) & !(c$Gene %in% b$Gene) ,]


write.table(a2, "~/Documents/NIEHS_LSU_UCD/niehs/multipop/190207_adjp005_unique_pop.csv", col.names=TRUE, row.names=FALSE, sep=",")
write.table(b2, "~/Documents/NIEHS_LSU_UCD/niehs/multipop/190207_adjp005_unique_exposure.csv", col.names=TRUE, row.names=FALSE, sep=",")
write.table(c2, "~/Documents/NIEHS_LSU_UCD/niehs/multipop/190207_adjp005_unique_pop_exposure.csv", col.names=TRUE, row.names=FALSE, sep=",")

```

Check that subsetting worked: 

```{r}
# p <- read.csv("~/Documents/NIEHS_LSU_UCD/niehs/multipop/190207_adjp005_unique_pop.csv")
# e <- read.csv("~/Documents/NIEHS_LSU_UCD/niehs/multipop/190207_adjp005_unique_exposure.csv")
# pe <- read.csv("~/Documents/NIEHS_LSU_UCD/niehs/multipop/190207_adjp005_unique_pop_exposure.csv")

intersect(a2$Gene, c2$Gene)

```

### Set up gene expression tables for MeV Heatmap

Extract read counts by DEG. Make gene lists for each treatment effect: 

```{r}

all_cof <- list("pop"=ttrans[,as.vector(a2$Gene)], "exposure"= ttrans[,as.vector(b2$Gene)], "pop_exposure"= ttrans[,as.vector(c2$Gene)])

trtnorm <- list()
gnorm <- list()
tmv <- list()
for (i in 1:length(all_cof)) {
  tmv[[i]] <- as.data.frame(all_cof[[i]])
  tmv[[i]]$trt <- y$samples$trt
  tmv[[i]] <- MeanVar(tmv[[i]])
  tmv[[i]] <- t(tmv[[i]]$mean)
  #tmv[[i]] <- tmv[[i]][,sort(colnames(tmv[[i]]))]
  tmv[[i]] <- tmv[[i]][,c("ARS.00", "ARS.10", "ARS.32", "ARS.56", 
                          "GT.00", "GT.10", "GT.32", "GT.56",
                          "GB.00", "GB.10", "GB.32", "GB.56",
                          "VB.00", "VB.10", "VB.32", "VB.56")]
  trtnorm[[i]] <- as.data.frame(cbind(tmv[[i]][,1:4]-tmv[[i]][,1], tmv[[i]][,5:8]-tmv[[i]][,5], tmv[[i]][,9:12]-tmv[[i]][,9], tmv[[i]][,13:16]-tmv[[i]][,13]))
  gnorm[[i]] <- as.data.frame(tmv[[i]]-rowMeans(tmv[[i]]))
  write.table(trtnorm[[i]], paste("~/Documents/NIEHS_LSU_UCD/niehs/multipop/190207_", names(all_cof[i]), "_trtnorm.txt", sep=""), col.names=NA, row.names=TRUE, sep="\t")
  write.table(gnorm[[i]], paste("~/Documents/NIEHS_LSU_UCD/niehs/multipop/190207_", names(all_cof[i]), "_gnorm.txt", sep=""), col.names=NA, row.names=TRUE, sep="\t")
}

```


## DAVID Analysis set up
After generating heatmaps and performing cluster analysis, save all clustered genes into a multi-list file for DAVID, making sure to extract only the genes with uniprot accession numbers

1. List saved cluster files and import their data
2. Make empty dataframe with n columns, n=number of total cluster files
3. Name of cluster file --> name of column, pipe funhe gene id's from each file into each column
4. Make another empty dataframe with same column names and number of columns
5. For every item in the column vector, if there's an entry for the uniprot accession, then write that uniprot entry into the dataframe
6. Repeat above for background genelists for each stage.

```{r}
dir <- "~/Documents/NIEHS_LSU_UCD/niehs/multipop"
files_list <- list.files(path = dir, pattern = "190219_")
files <- file.path(dir, files_list)
filenames <- gsub("190219_", "", files_list)
filenames <- gsub(".txt", "", filenames)

#Dump cluster gene lists into a list element
geneclusters <- list()
for (i in 1:length(files)) {
  geneclusters[[i]] <- read.csv(files[i], header=TRUE, sep = "\t")
  colnames(geneclusters[[i]])[2] <- "Gene"
  names(geneclusters)[i] <- filenames[i]
  geneclusters[[i]] <- as.vector(geneclusters[[i]]$Gene)
}

#Filter above genes by those with uniprot accession number: 
library(readxl)
genemodels <- as.data.frame(read_xlsx("~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/Gene.Models.Conservative.and.Liberal.xlsx"))
names(genemodels) <- c("geneID", "name", "cons", "lib")
unip <- list()

for (i in 1:length(geneclusters)){
  unip[[i]] <- genemodels[genemodels$geneID %in% geneclusters[[i]],]$lib
  unip[[i]] <- unip[[i]][!is.na(unip[[i]])]
  names(unip)[i] <- names(geneclusters)[i]
}
#Assign same length to all vectors, then run cbind: 
n <- max(lengths(unip))
for (i in 1:length(unip)) {
  length(unip[[i]]) <- n
}
david <- do.call(cbind, unip)
david[is.na(david)] <- " " 
write.table(david, file="~/Documents/NIEHS_LSU_UCD/niehs/multipop/190219_clustergenelists.txt", row.names = FALSE, col.names = TRUE, sep="\t", quote=FALSE)

#Compile background genelist for all-stages analysis by uniprot accession IDs and store one multi-list file: 

bg <- as.vector(rownames(y$counts))
bg_unip <- genemodels[genemodels$geneID %in% bg,]$lib
bg_unip <- bg_unip[!is.na(bg_unip)]
bg_unip <- as.data.frame(bg_unip)
write.table(bg_unip, file="~/Documents/NIEHS_LSU_UCD/niehs/multipop/190219_backgroundgenelists.txt", row.names=FALSE, col.names=TRUE, sep="\t", quote=FALSE)

```


## Venn Diagrams: 

```{r}

##Venn Diagram of parent DEGs across stage-specific analyses: 
library(VennDiagram)
venn.diagram(
x = list(e_g$Gene, m_g$Gene, l_g$Gene),
category.names = c("Stage 19 Parent DEG" , "Stage 28 Parent DEG " , "Stage 35 Parent DEG"),
filename = 'parent_DEG_venn.tiff',
        output = TRUE ,
        height = 3000, 
        width = 3000, 
        resolution = 500, 
        imagetype = "tiff", 
        units = "px", 
        compression ="lzw",
        lwd = 2,
        lty = 'blank',
        fill = c('yellow', 'purple', 'green'),
        cex = 1,
        fontface = "bold",
        fontfamily = "sans",
        cat.cex = 0.6,
        cat.fontface = "bold",
        cat.default.pos = "outer",
        cat.pos = c(-27, 27, 135),
        cat.dist = c(0.055, 0.055, 0.085),
        cat.fontfamily = "sans",
        rotation = 1)


##Venn Diagram of exposure genes across stages: 

venn.diagram(
x = list(e_ex$Gene, m_ex$Gene, l_ex$Gene),
category.names = c("Stage 19 Embryo Exp DEG" , "Stage 28 Embryo Exp DEG " , "Stage 35 Embryo Exp DEG"),
filename = 'exp_DEG_venn.tiff',
        output = TRUE ,
        height = 3000, 
        width = 3000, 
        resolution = 500, 
        imagetype = "tiff", 
        units = "px", 
        compression ="lzw",
        lwd = 2,
        lty = 'blank',
        fill = c('yellow', 'purple', 'green'),
        cex = 1,
        fontface = "bold",
        fontfamily = "sans",
        cat.cex = 0.6,
        cat.fontface = "bold",
        cat.default.pos = "outer",
        cat.pos = c(-27, 27, 135),
        cat.dist = c(0.055, 0.055, 0.085),
        cat.fontfamily = "sans",
        rotation = 1)

```

## Calculate mean-normalized expression values for the DEG for parental main effect at Stage 19 across all stages: 

```{r}
#Import gene order from MeV hierarchical clustering: 
gorder <- read.csv("~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180403_early_group_geneorder.txt", header=TRUE, sep = "\t")

#Make sure that the expression values pulled from larger data set look right by looking at expression patterns in just stage 19 samples from larger dataset y: 
egtest <- y[as.vector(gorder$X), y$samples$stage=="19"]$counts

tmv <- as.data.frame(t(egtest))
tmv <- TreatNames_sub(tmv)
tmv <- MeanVar(tmv)
tmv <- t(tmv$mean)
tmv <- tmv[,sort(colnames(tmv))]
tmv <- tmv - rowMeans(tmv)
write.table(tmv, "~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180329_st19_test.txt", col.names=NA, row.names=TRUE, sep="\t")


#Extract expression values for all stages from the larger data set: 
early_g <- y[as.vector(gorder$X), y$samples$stage=="19"]$counts
mid_g <- y[as.vector(gorder$X), y$samples$stage=="28"]$counts
late_g <- y[as.vector(gorder$X), y$samples$stage=="35"]$counts

eg <- list(early_g, mid_g, late_g)

#Normalize to each stage mean for visualization: 
for (i in 1:length(eg)) {
  eg[[i]] <- as.data.frame(t(eg[[i]]))
  eg[[i]] <- TreatNames_sub(eg[[i]])
  eg[[i]] <- MeanVar(eg[[i]])
  eg[[i]] <- t(eg[[i]]$mean)
  eg[[i]] <- eg[[i]][,sort(colnames(eg[[i]]))]
  eg[[i]] <- eg[[i]] - rowMeans(eg[[i]])
}

eg <- do.call(cbind, eg)
write.table(eg, "~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180403_st19_parentDEG_allstages.txt", col.names=NA, row.names=TRUE, sep="\t")

#Normalize to the total grand mean: 
# all_deg <- y[as.vector(gorder$X), ]$counts
# tmv <- as.data.frame(t(all_deg))
# tmv <- TreatNames_sub(tmv)
# tmv <- MeanVar(tmv)
# tmv <- t(tmv$mean)
# tmv <- tmv[,sort(colnames(tmv))]
# tmv <- tmv - rowMeans(tmv)
# write.table(tmv, "~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180314_st19_parentDEG_allgnorm.txt", col.names=NA, row.names=TRUE, sep="\t")


```


#### Visualizing parental main effect genes from mid and late stage-specific analyses across whole data set: 

```{r}
#Import gene order from hierarchical clustering from MeV
m_g <- read.csv("~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180403_st28_group_geneorder.txt", header=TRUE, sep="\t")
l_g <- read.csv("~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180403_st35_group_geneorder.txt", header=TRUE, sep = "\t")


#Normalize expression values to the grand mean for each stage:
mg_19 <- y[as.vector(m_g$X), y$samples$stage=="19"]$counts
mg_28 <- y[as.vector(m_g$X), y$samples$stage=="28"]$counts
mg_35 <- y[as.vector(m_g$X), y$samples$stage=="35"]$counts

lg_19 <- y[as.vector(l_g$X), y$samples$stage=="19"]$counts
lg_28 <- y[as.vector(l_g$X), y$samples$stage=="28"]$counts
lg_35 <- y[as.vector(l_g$X), y$samples$stage=="35"]$counts

mg <- list(mg_19, mg_28, mg_35)
lg <- list(lg_19, lg_28, lg_35)

for (i in 1:length(mg)) {
  mg[[i]] <- as.data.frame(t(mg[[i]]))
  mg[[i]] <- TreatNames_sub(mg[[i]])
  mg[[i]] <- MeanVar(mg[[i]])
  mg[[i]] <- t(mg[[i]]$mean)
  mg[[i]] <- mg[[i]][,sort(colnames(mg[[i]]))]
  mg[[i]] <- mg[[i]] - rowMeans(mg[[i]])
}
mg <- do.call(cbind, mg)

for (i in 1:length(lg)) {
  lg[[i]] <- as.data.frame(t(lg[[i]]))
  lg[[i]] <- TreatNames_sub(lg[[i]])
  lg[[i]] <- MeanVar(lg[[i]])
  lg[[i]] <- t(lg[[i]]$mean)
  lg[[i]] <- lg[[i]][,sort(colnames(lg[[i]]))]
  lg[[i]] <- lg[[i]] - rowMeans(lg[[i]])
}
lg <- do.call(cbind, lg)

write.table(mg, "~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180404_st28_parent_allstages.txt", col.names=NA, row.names=TRUE, sep="\t")
write.table(lg, "~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180404_st35_parent_allstages.txt", col.names=NA, row.names=TRUE, sep="\t")

```


## Look at inter-individual variation in non-DEG 

Visualize gene expression between parent control and parent exposed groups, or plot inviduals on MDS space:

```{r}

#Drop all DEG from stage 19 dataframe: 
e_g <- read.csv("~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180329_adjp005_ear_unique_group.csv")
e_ex <- read.csv("~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180329_adjp005_ear_unique_exposure.csv")
e_ge <- read.csv("~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180329_adjp005_ear_unique_gxe.csv")

exgenes <- c(as.vector(e_g$Gene), as.vector(e_ex$Gene), as.vector(e_ge$Gene))
ex <- which(exgenes %in% rownames(ear$counts))
mat <- ear[-ex,]

#Make MDS plot:
par(mfrow=c(1,1))
group <- mat$samples$group
treats <- mat$samples$trt
exp <- mat$samples$exposure

col.group <- as.factor(group)
levels(col.group) <- c("blue", "red")
col.group <- as.character(col.group)

col.exp <- as.factor(exp)
levels(col.exp) <- brewer.pal(nlevels(col.exp), "Dark2")
col.exp <- as.character(col.exp)

col.treat <- as.factor(treats)
levels(col.treat) <- brewer.pal(nlevels(col.treat), "Set1")
col.treat <- as.character(col.treat)

plotMDS(mat, top=19974, labels=treats, col=col.group, dim=c(1,2), gene.selection="common", method="bcv", main="Stage 19 Individuals Expression of Background Genes")


#Make PCA Plot: 

library("FactoMineR")
library("factoextra")

#Transpose data frame to put variables(genes) into columns: 
matc <- t(mat$counts)

#Compute PCA:
res.pca <- prcomp(matc, scale=TRUE)

#Visualize eigenvalues
fviz_eig(res.pca)

#Visualize individuals in PCA plot: 
groups <- as.factor(mat$samples$group)
doses <- as.factor(mat$samples$exposure)

e_pca <- fviz_pca_ind(res.pca,
             #geom.ind = "point",
             col.ind = groups, # Color by groups
             palette = c("#00AFBB","#FC4E07"),
             addEllipses = TRUE,
             ellipse.type = "confidence",
             legend.title = "Parent Treatment Group",
             repel = FALSE) +    # Avoid text overlapping 
             labs(title ="Stage 19 Individuals")

#Look at dimension description:
ind <- get_pca_ind(res.pca)
ind
res.desc <- dimdesc(res.pca, axes=1:2, proba=0.05)

fviz_contrib(res.pca, choice = "ind", axes = 1:2)

```


```{r}
m <- c(as.vector(m_g$Gene), as.vector(m_ex$Gene), as.vector(m_ge$Gene))
l <- c(as.vector(l_g$Gene), as.vector(l_ex$Gene), as.vector(l_ge$Gene))

mex <- which(m %in% rownames(mid$counts)) 
lex <- which(l %in% rownames(lat$counts))

mpca <- t(mid$counts[-mex,])
lpca <- t(lat$counts[-lex,])

#Compute PCA:
m.pca <- prcomp(mpca, scale=TRUE)
l.pca <- prcomp(lpca, scale=TRUE)

#Visualize eigenvalues
fviz_eig(m.pca)
fviz_eig(l.pca)

#Visualize individuals in PCA plot: 
m_groups <- as.factor(mid$samples$group)
m_doses <- as.factor(mid$samples$exposure)

l_groups <- as.factor(lat$samples$group)
l_doses <- as.factor(lat$samples$exposure)

m_pca <- fviz_pca_ind(m.pca,
             #geom.ind = "point",
             col.ind = m_groups, # Color by groups
             palette = c("#00AFBB","#FC4E07"),
             addEllipses = TRUE,
             ellipse.type = "confidence",
             legend.title = "Parent Treatment",
             repel = FALSE) +
   labs(title ="Stage 28 Individuals")

l_pca <- fviz_pca_ind(l.pca,
             #geom.ind = "point",
             col.ind = l_groups, # Color by groups
             palette = c("#00AFBB","#FC4E07"),
             addEllipses = TRUE,
             ellipse.type = "confidence",
             legend.title = "Parent Treatment",
             repel = FALSE)   +
            labs(title ="Stage 35 Individuals")

all_pca <- multiplot(e_pca, m_pca, l_pca, cols=3)

```

## Look at inter-individual variation in DEG expression between parent control and parent exposed groups, or plot inviduals on MDS space:

```{r}

#Drop all DEG from stage 19 dataframe: 
e_g <- read.csv("~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180329_adjp005_ear_unique_group.csv")
m_g <- read.csv("~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180329_adjp005_mid_unique_group.csv")
l_g <- read.csv("~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180329_adjp005_lat_unique_group.csv")

all_g <- list("Early"=e_g, "Mid"=m_g, "Late"=l_g)
ex <- vector("list", 3)
for (i in 1:3) {
  all_g[[i]] <- as.vector(all_g[[i]]$Gene)
  ex[[i]] <- which(rownames(StageData[[i]]$counts) %in% ex[[i]])
  all_g[[i]] <- StageData[[i]][all_g[[i]],]
}

#Make PCA Plot: 

library("FactoMineR")
library("factoextra")

#Transpose data frame to put variables(genes) into columns: 
mat <- vector("list", 3)
for (i in 1:3) {
  mat[[i]] <- t(all_g[[i]]$counts)
}

#Compute PCA:
res.pca <- lapply(mat, "prcomp")

#Visualize eigenvalues
#fviz_eig(res.pca)

#Visualize individuals in PCA plot: 

doses <- as.factor(y$samples$exposure)
pops <- as.factor(y$samples$pop)


edeg_pca <- fviz_pca_ind(res.pca[[1]],
             #geom.ind = "point",
             col.ind = e_groups, # Color by groups
             palette = c("#00AFBB","#FC4E07"),
             addEllipses = TRUE,
             ellipse.type = "confidence",
             legend.title = "Parent Treatment Group",
             repel = FALSE) +    # Avoid text overlapping 
             labs(title ="Stage 19 Individuals")

mdeg_pca <- fviz_pca_ind(res.pca[[2]],
             #geom.ind = "point",
             col.ind = m_groups, # Color by groups
             palette = c("#00AFBB","#FC4E07"),
             addEllipses = TRUE,
             ellipse.type = "confidence",
             legend.title = "Parent Treatment Group",
             repel = FALSE) +    # Avoid text overlapping 
             labs(title ="Stage 28 Individuals")

ldeg_pca <- fviz_pca_ind(res.pca[[3]],
             #geom.ind = "point",
             col.ind = l_groups, # Color by groups
             palette = c("#00AFBB","#FC4E07"),
             addEllipses = TRUE,
             ellipse.type = "confidence",
             legend.title = "Parent Treatment Group",
             repel = FALSE) +    # Avoid text overlapping 
             labs(title ="Stage 35 Individuals")

edeg_pca
mdeg_pca
ldeg_pca
gdeg_pca <- multiplot(edeg_pca, mdeg_pca, ldeg_pca, cols=3)

#Make MDS plot:
par(mfrow=c(1,1))
group <- mat$samples$group
treats <- mat$samples$trt
exp <- mat$samples$exposure

col.group <- as.factor(group)
levels(col.group) <- c("blue", "red")
col.group <- as.character(col.group)

col.exp <- as.factor(exp)
levels(col.exp) <- brewer.pal(nlevels(col.exp), "Dark2")
col.exp <- as.character(col.exp)

col.treat <- as.factor(treats)
levels(col.treat) <- brewer.pal(nlevels(col.treat), "Set1")
col.treat <- as.character(col.treat)

plotMDS(mat, labels=treats, col=col.group, main="Stage 19 Individuals By Parent DEGs")

```


## Comparing the variance of non-DEG expression between Exposed and Control Parent Groups

For stage 19: 

```{r}
#Using all genes, including DEG: 
pc <- ear[,ear$samples$group=="C"]
pe <- ear[, ear$samples$group=="E"]


# #Drop all DEG from stage 19 dataframe, also subsetting by parent treatment group: 
# e_g <- read.csv("~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180329_adjp005_ear_unique_group.csv")
# e_ex <- read.csv("~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180329_adjp005_ear_unique_exposure.csv")
# e_ge <- read.csv("~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180329_adjp005_ear_unique_gxe.csv")
# 
# exgenes <- c(as.vector(e_g$Gene), as.vector(e_ex$Gene), as.vector(e_ge$Gene))
# ex <- which(exgenes %in% rownames(ear$counts))
# 
# pc <- ear[-ex, ear$samples$group=="C"]
# pe <- ear[-ex, ear$samples$group=="E"]

#Calculate variance for each treatment group and visualize in density plot: 
pc  <- as.data.frame(t(pc$counts))
pc <- TreatNames_sub(pc)
pc <- MeanVar(pc)

pe  <- as.data.frame(t(pe$counts))
pe <- TreatNames_sub(pe)
pe <- MeanVar(pe)

#Combine variances from each data frame into one data frame for visualization: 
pc <- pc$stdev
pe <- pe$stdev

library(tidyr)
library(dplyr)
library(data.table)
library(ggplot2)
pc1 <- setDT(as.data.frame(pc), keep.rownames = TRUE)

pc1 <- gather(pc1, Gene, stdev, colnames(pc1)[2:ncol(pc1)])
pc1$group <- 'C'

pc2 <- setDT(as.data.frame(pe), keep.rownames = TRUE)
pc2 <- gather(pc2, Gene, stdev, colnames(pc2)[2:ncol(pc2)])
pc2$group <- 'E'

#Now, combine your two dataframes into one.  First make a new column in each that will be a variable to identify where they came from later.

early_var <- as.data.frame(rbind(pc1, pc2))

#After that, which is unnecessary if your data is in long formal already, you only need one line to make your plot.

stdens <- ggplot(early_var, aes(stdev, fill = group)) +
  geom_density(alpha = 0.2) +
  scale_fill_brewer(palette="Dark2") +
  theme_classic() + 
  ggtitle("Density of Variance of All Non-DEG at Stage 19")

stdens

```

Repeat the process above for stage 28 and 35 genes: 

```{r}

#Load all the DEGs found at each stage
m_g <- read.csv("~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180329_adjp005_mid_unique_group.csv")
m_ex <- read.csv("~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180329_adjp005_mid_unique_exposure.csv")
m_ge <- read.csv("~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180329_adjp005_mid_unique_gxe.csv")

l_g <- read.csv("~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180329_adjp005_lat_unique_group.csv")
l_ex <- read.csv("~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180329_adjp005_lat_unique_exposure.csv")
l_ge <- read.csv("~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180329_adjp005_lat_unique_gxe.csv")

#Set up your data frames to subset by stage and non-DEG
m <- c(as.vector(m_g$Gene), as.vector(m_ex$Gene), as.vector(m_ge$Gene))
l <- c(as.vector(l_g$Gene), as.vector(l_ex$Gene), as.vector(l_ge$Gene))

mex <- which(m %in% rownames(mid$counts))
lex <- which(l %in% rownames(lat$counts))

mc <- mid[-mex, mid$samples$group=="C"]
me <- mid[-mex, mid$samples$group=="E"]
m28 <- list(mc, me)

lc <- lat[-lex, lat$samples$group=="C"]
le <- lat[-lex, lat$samples$group=="E"]
l35 <- list(lc, le)

ml <- list("mc"=mc, "me"=me, "lc"=lc, "le"=le)

#Calculate variance for each treatment group and visualize in histogram: 
ml_var <- list()

for (i in 1:length(ml)) {
  ml[[i]] <- as.data.frame(t(ml[[i]]$counts))
  ml[[i]] <- TreatNames_sub(ml[[i]])
  ml[[i]] <- MeanVar(ml[[i]])
  ml[[i]] <- ml[[i]]$stdev
  ml_var[[i]] <- setDT(as.data.frame(ml[[i]]), keep.rownames = TRUE)
  ml_var[[i]] <- gather(ml_var[[i]], Gene, stdev, colnames(ml_var[[i]])[2:ncol(ml_var[[i]])])
}


m_pc <- ml_var[[1]]
m_pc$group <- 'C'
m_pe <- ml_var[[2]]
m_pe$group <- 'E'
mid_var <- rbind(m_pc, m_pe)

mvar <- ggplot(mid_var, aes(stdev, fill = group)) +
  geom_density(alpha = 0.2) +
  scale_fill_brewer(palette="Dark2") +
  theme_classic() +
  ggtitle("Density of Variance of All Non-DEG at Stage 28")

l_pc <- ml_var[[3]]
l_pc$group <- 'C'
l_pe <- ml_var[[4]]
l_pe$group <- 'E'
late_var <- rbind(l_pc, l_pe)

lvar <- ggplot(late_var, aes(stdev, fill = group)) +
  geom_density(alpha = 0.2) +
  scale_fill_brewer(palette="Dark2") +
  theme_classic() +
  ggtitle("Density of Variance of All Non-DEG at Stage 35")

all_var <- multiplot(stdens, mvar, lvar, col=1)

```


## Visualize the log2FC of Stage19 Parent Group DEGs across stages

```{r}
library(data.table)
source("~/Documents/NIEHS_LSU_UCD/scripts/fun_multiplot.R")

e_g <- read.csv("~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180329_adjp005_ear_unique_group.csv")
m_g <- read.csv("~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180329_adjp005_mid_unique_group.csv")
l_g <- read.csv("~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180329_adjp005_lat_unique_group.csv")
all_g <- list("e"=e_g, "m"=m_g, "l"=l_g)

#Subset larger data set by the DEGs for parent main effect from each stage analysis, and calculate logFC between pe and pc treatments: 
degs <- list()
for (i in 1:length(all_g)) {
  degs[[i]] <- y[as.vector(all_g[[i]]$Gene),]
  degs[[i]] <- as.data.frame(t(degs[[i]]$counts))
  degs[[i]] <- TreatNames_sub(degs[[i]])
  degs[[i]] <- TreatMeans(degs[[i]])
  degs[[i]] <- t(degs[[i]])
  degs[[i]] <- degs[[i]][,sort(colnames(degs[[i]]))]
  
  degs[[i]] <- data.frame(cbind("early" = rowMeans(degs[[i]][,13:16]) - rowMeans(degs[[i]][,1:4]), 
              "mid" = rowMeans(degs[[i]][,17:20]) - rowMeans(degs[[i]][,5:8]), 
              "late"= rowMeans(degs[[i]][,21:24]) - rowMeans(degs[[i]][,9:12])))
  
  degs[[i]] <- setDT(degs[[i]], keep.rownames = TRUE)
  colnames(degs[[i]])[1] <- "Gene"
  degs[[i]]$DEG <- "DE"
  degs[[i]] <- gather(degs[[i]], Stage, logFC, "early":"late")
  degs[[i]]$absFC <- abs(degs[[i]]$logFC)
}

names(degs) <- c("Stage 19", "Stage 28", "Stage 35")

#Make graph:
hoc_plots <- list()
for (i in 1:length(degs)) {
  hoc_plots[[i]] <- ggplot(data=degs[[i]], aes(x=Stage, y=logFC, group=Gene)) +
    geom_line() +
    geom_point() +
    scale_x_discrete(limits=c("early","mid","late")) +
    xlab("Developmental Stage") +
    ggtitle(names(degs)[[i]])
}

hoc_plots[[2]]
hcplots_all <- multiplot(hoc_plots[[1]], hoc_plots[[2]], hoc_plots[[3]], cols=3)
```


## Visualize distribution of logFC between PC and PE of Stage 19 Parent DEGs across development

```{r}
#First establish dataframe of logFC of randomly sampled genes from larger data set: 
testrand <- list()
ty <- list()
fc_dens <- list()

for (i in 1:length(all_g)) {
  testrand[[i]] <- replicate(1000, sample(rownames(y$counts), size=nrow(all_g[[i]]), replace=FALSE))
}

for (i in 1:length(testrand)) {
  for (j in 1:ncol(testrand[[i]])) {
    ty[[j]] <- y[as.vector(testrand[[i]][,j]),]$counts
    ty[[j]]  <- as.data.frame(t(ty[[j]]))
    ty[[j]] <- TreatNames_sub(ty[[j]])
    ty[[j]] <- TreatMeans(ty[[j]])
    ty[[j]] <- t(ty[[j]])
    ty[[j]] <- data.frame(ty[[j]][,sort(colnames(ty[[j]]))])
    
    ty[[j]] <- data.frame(cbind("early" = rowMeans(ty[[j]][,13:16]) - rowMeans(ty[[j]][,1:4]), 
                                "mid" = rowMeans(ty[[j]][,17:20]) - rowMeans(ty[[j]][,5:8]), 
                                "late"= rowMeans(ty[[j]][,21:24]) - rowMeans(ty[[j]][,9:12])))
    
    ty[[j]] <- setDT(ty[[j]], keep.rownames = TRUE)
    colnames(ty[[j]])[1] <- "Gene"
    ty[[j]]$DEG <- j 
    ty[[j]] <- gather(ty[[j]], Stage, logFC, "early":"late")
    fc_dens[[i]] <- rbindlist(ty)
  }
  fc_dens[[i]]$absFC <- abs(fc_dens[[i]]$logFC)
  fc_dens[[i]] <- rbind(fc_dens[[i]], degs[[i]])
}

names(fc_dens) <- names(degs)

for (i in 1:3) {
  fc_dens[[i]]$Stage <- factor(fc_dens[[i]]$Stage, levels=c("early", "mid", "late"))
}


#Visualize distribution of the absolute value logFC of 245 Stage 19 parent DEGs vs 245 randomly selected genes at Stage 19: 
st19p <- ggplot(fc_dens[[1]], aes(x = absFC, fill=DEG=="DE")) +
  facet_wrap(~Stage, as.table=FALSE, dir="v") + 
  geom_density(aes(y=..scaled..), alpha = 0.4) +
  scale_fill_manual(values= c("FALSE"="grey", "TRUE"="Green")) + 
  theme_classic() + 
  xlab("absolute(logFC)") +
  ylab("Density") +
  labs(title=paste('Distribution of', names(fc_dens)[1], 'Parent DEG LogFC', sep=" "), subtitle="245 DEG") 

st28p <- ggplot(fc_dens[[2]], aes(x = absFC, fill=DEG=="DE")) +
  facet_wrap(~Stage, as.table=FALSE, dir="v") + 
  geom_density(aes(y=..scaled..), alpha = 0.4) +
  scale_fill_manual(values= c("FALSE"="grey", "TRUE"="#00CC99")) + 
  theme_classic() + 
  xlab("absolute(logFC)") +
  ylab("Density") +
  labs(title=paste('Distribution of', names(fc_dens)[2], 'Parent DEG LogFC', sep=" "), subtitle="211 DEG")

st35p <- ggplot(fc_dens[[3]], aes(x = absFC, fill=DEG=="DE")) +
  facet_wrap(~Stage, as.table=FALSE, dir="v") + 
  geom_density(aes(y=..scaled..), alpha = 0.4) +
  scale_fill_manual(values= c("FALSE"="grey", "TRUE"="#006600")) + 
  theme_classic() + 
  xlab("absolute(logFC)") +
  ylab("Density") +
  labs(title=paste('Distribution of', names(fc_dens)[3], 'Parent DEG LogFC', sep=" "), subtitle="566 DEG")

fcplots_all <- multiplot(st19p, st28p, st35p, cols=3)


#Alternative to facet_wrap
densplots <- vector("list", 3) 
for (j in 1:3) {
  densplots[[1]][[j]] <- ggplot(fc_dens[[1]][fc_dens[[1]]$Stage==unique(fc_dens[[1]]$Stage)[j]], aes(x = absFC, fill=DEG=="DE")) +
  geom_density(aes(y=..scaled..), alpha = 0.4) +
  scale_fill_manual(values= c("FALSE"="grey", "TRUE"="Green")) + 
  theme_classic() + 
  xlab("absolute(logFC)") +
  ylab("Density") +
  ggtitle(paste('Distribution of', names(fc_dens)[1], 'Parent DEG LogFC in', names(fc_dens)[j], sep=" ")) 
  
  densplots[[2]][[j]] <- ggplot(fc_dens[[2]][fc_dens[[2]]$Stage==unique(fc_dens[[2]]$Stage)[j]], aes(x = absFC, fill=DEG=="DE")) +
  geom_density(aes(y=..scaled..), alpha = 0.4) +
  scale_fill_manual(values= c("FALSE"="grey", "TRUE"="#00CC99")) + 
  theme_classic() + 
  xlab("absolute(logFC)") +
  ylab("Density") +
  ggtitle(paste('Distribution of', names(fc_dens)[2], 'Parent DEG LogFC in', names(fc_dens)[j], sep=" ")) 
  
  densplots[[3]][[j]] <- ggplot(fc_dens[[3]][fc_dens[[3]]$Stage==unique(fc_dens[[3]]$Stage)[j]], aes(x = absFC, fill=DEG=="DE")) +
  geom_density(aes(y=..scaled..), alpha = 0.4) +
  scale_fill_manual(values= c("FALSE"="grey", "TRUE"="#006600")) + 
  theme_classic() + 
  xlab("absolute(logFC)") +
  ylab("Density") +
  ggtitle(paste('Distribution of', names(fc_dens)[3], 'Parent DEG LogFC in', names(fc_dens)[j], sep=" ")) 
  
}

fcplots_all <- multiplot(densplots[[1]][[1]], densplots[[1]][[2]], densplots[[1]][[3]], 
                         densplots[[2]][[1]], densplots[[2]][[2]], densplots[[2]][[3]],
                         densplots[[3]][[1]], densplots[[3]][[2]], densplots[[3]][[3]], cols=3)




```


## Control Embryo-specific analyses: 

Subset just the 00% embryo exposure samples for control embryo-specific analyses: 

```{r}
ce <- zz[,zz$samples$exposure=="00"]
```

#### Filter control embryo samples: 
```{r}

#Create list of genes with more than 5 counts for at least 4 samples in each treatment group: 
d = list()
treat <- unique(ce$samples$trt)
head(treat)
length(treat)

for (i in 1:length(treat)) {
	a <- which(ce$samples$trt==treat[i])
	b <- colnames(ce$counts[,a])
	kp <- rowSums(ce$counts[,a] > 5) >=4 
	g <- which(kp==TRUE)
	d[[i]] <- g
}

#Subset control embryo subset by the genes identified above: 
nrow(ce$counts)
filt_genes <- unique(names(unlist(d, use.names=TRUE)))
length(filt_genes)
ce <- ce[filt_genes,]
nrow(ce$counts)

#Save filtered control embryo background genes:
write.table(filt_genes, file="~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180322_controlemb_no_outliers_backgroundlist_noouts.csv",
	col.names=FALSE, row.names=FALSE, sep="\t")

#Transform the control embryo subset by the same constant: 
ce_transform <- log2(ce$counts +8)

#Repeat the above to check that transformation is good with control embryo-specific subset (because there are far fewer samples in the set)
ce_ttrans <- t(ce_transform)
ce_ttrans <- as.data.frame(ce_ttrans)
ce_ttrans <- TreatNames_sub(ce_ttrans)
mv <- MeanVar(ce_ttrans)
mea <- as.vector(mv$mean)
sd <- as.vector(mv$stdev)

slope <- (coef(lm(sd~mea))[2])
print(slope)

#Normalization of control embryo subset: 
ce_mean_counts <- colMeans(ce_transform)
ce_norm_counts <- sweep(ce_transform, 2, colMeans(ce_transform))
ce_grandmean <- mean(ce_mean_counts)
ce_norm_counts <- ce_norm_counts + ce_grandmean

#Save normalized counts and samples for control embryo subset: 
ce_norm_counts <- as.data.frame(ce_norm_counts)
write.table(ce_norm_counts, "~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180305_controlemb_no_outlier_normcounts_noouts.csv", 
		col.names=TRUE, row.names=TRUE, sep=",")
write.table(ce$samples, "~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180305_controlemb_no_outlier_samplesinfo_noouts.csv", 
            col.names=TRUE, row.names=TRUE, sep=",") 
```

## Test for differential expression

#### Control embryo subset
```{r}
#Format the dataframe to run lm() for every gene:
ce$counts <- ce_norm_counts
ce.deg <- t(ce$counts)
ce.deg <- as.data.frame(ce.deg)
ce.deg <- TreatNames_sub(ce.deg)

#Test lm() on small subset of genes:
t <- dim(ce.deg)[2]
test <- ce.deg[,(t-10):t]
l <- lm(formula=test[,"Funhe2EKm042398"] ~ group*stage, data=test)
a <- anova(l)

#Set up data frame for storing p-values from lm():
g <- dim(ce.deg)[2]-5
genelist.ce <- colnames(ce.deg[,1:g])
cof.ce <- rownames(a)[1:3]
pdat.ce <- matrix(nrow=length(genelist.ce), ncol=length(cof.ce))
colnames(pdat.ce) <- cof.ce
rownames(pdat.ce) <- genelist.ce

#Run lm() for every gene:
for (i in 1:ncol(ce.deg[,1:g])) {
	g <- which(rownames(pdat.ce)==colnames(ce.deg)[i])
	l <- lm(formula=ce.deg[,i] ~ group*stage, data=ce.deg)
	a <- anova(l)
	p <- a$`Pr(>F)`[1:3]
	pdat.ce[g,] <- p 
}

#Save p-values
pdat.ce <- as.data.frame(pdat.ce)
names(pdat.ce) <- gsub(":", "_", names(pdat.ce))
write.table(pdat.ce, "~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180322_exp00_anova_unadj_pvalues_noouts.csv", 
		col.names=TRUE, row.names=TRUE, sep=",")

#Adjust for FDR:
fdr.ce <- data.frame(pdat.ce)
adj.fdr.ce <- setDT(fdr.ce, keep.rownames = TRUE)
colnames(adj.fdr.ce)[1] <- "Gene"
adj.fdr.ce <- adj.fdr.ce %>%
  gather(`group`, `stage`, `group_stage`,
         key="cof", value="pvals")
adj.fdr.ce$adj.pvals <- p.adjust(adj.fdr.ce$pvals, method="fdr")
deg.ce <- adj.fdr.ce[adj.fdr.ce$adj.pvals <= 0.05,]

##For control embryo subset: 
write.table(adj.fdr.ce, "~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180305_all_fdradj_pval_exp00_noouts.csv", col.names=TRUE, row.names=FALSE, sep=",")
##For control embryo subset: 
write.table(deg.ce, "~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180305_fdradj_005_deg_exp00_noouts.csv", col.names=TRUE, row.names=FALSE, sep=",")

##Write each DEG by coefficient for control emb exposure samples: 
effects = list()
cof <- unique(deg.ce$cof)
for (i in 1:length(cof)) {
  a <- which(cof[i] == deg.ce$cof)
  effects[[cof[i]]] <- deg.ce[a,]
  write.table(effects[[cof[i]]], paste("~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180305_adjp005_exp00_",  names(effects[cof[i]]), "_noouts.csv", sep=""),
              col.names=TRUE, row.names=FALSE, sep=",")
}

##Repeat for control emb exposure samples: 
a <- effects$group
b <- effects$stage
c <- effects$group_stage
a2 <- a[!(a$Gene %in% c$Gene) & !(a$Gene %in% b$Gene),]
b2 <- b[!(b$Gene %in% c$Gene) & !(b$Gene %in% a$Gene),]
write.table(a2, "~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180305_adjp005_exp00_unique_group_noouts.csv", col.names=TRUE, row.names=FALSE, sep=",")
write.table(b2, "~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180305_adjp005_exp00_unique_stage_noouts.csv", col.names=TRUE, row.names=FALSE, sep=",")
write.table(c, "~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180305_adjp005_exp00_unique_group_stage_noouts.csv", col.names=TRUE, row.names=FALSE, sep=",")

##Repeat for control emb exposure samples: 
ce.cof <- list("group"=ce[as.vector(a2$Gene),]$counts, "stage"=ce[as.vector(b2$Gene),]$counts, "group_stage"=ce[as.vector(c$Gene),]$counts)
trtnorm.ce <- list()
gnorm.ce <- list()
names(trtnorm.ce) <- names(ce.cof)
names(gnorm.ce) <- names(ce.cof)
tmv.ce <- list()
for (i in 1:length(ce.cof)) {
  tmv.ce[[i]] <- as.data.frame(t(ce.cof[[i]]))
  tmv.ce[[i]] <- TreatNames_sub(tmv.ce[[i]])
  tmv.ce[[i]] <- MeanVar(tmv.ce[[i]])
  tmv.ce[[i]] <- t(tmv.ce[[i]]$mean)
  tmv.ce[[i]] <- tmv.ce[[i]][,sort(colnames(tmv.ce[[i]]))]
  trtnorm.ce[[i]] <- as.data.frame(cbind(tmv.ce[[i]][,1:3]-tmv.ce[[i]][,1], tmv.ce[[i]][,4:6]-tmv.ce[[i]][,4]))
  gnorm.ce[[i]] <- as.data.frame(tmv.ce[[i]]-rowMeans(tmv.ce[[i]]))
  write.table(trtnorm.ce[[i]], paste("~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180305_exp00_", names(ce.cof[i]), "_trtnorm.txt", sep=""), col.names=NA, row.names=TRUE, sep="\t")
  write.table(gnorm.ce[[i]], paste("~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/180305_exp00_", names(ce.cof[i]), "_gnorm.txt", sep=""), col.names=NA, row.names=TRUE, sep="\t")
}

```

Repeat above for control emb exposure. 
Make background list of Uniprot accession ids: 

```{r}
library(readr)
bg_ce <- read_csv("Ranalysis/180305_controlemb_backgroundlist_noouts.csv", col_names = FALSE)
bg_ce <- as.vector(bg_ce$X1)
bg_ce_unip <- genemodels[genemodels$geneID %in% bg_ce,]$lib
bg_ce_unip <- bg_ce_unip[!is.na(bg_ce_unip)]
write.table(bg_ce_unip, file="~/Documents/NIEHS_LSU_UCD/rnaseq/david_results/180305_exp00_backgroundgenelist.txt", row.names=FALSE, col.names=FALSE, sep="\t", quote=FALSE)

```

Make cluster gene lists of Uniprot accession ids for control embryos:

```{r}
dir <- "~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/mev_heatmap_expression_tables"
files_list <- list.files(path = dir, pattern = "180306_")
files <- file.path(dir, files_list)
filenames <- gsub("180306_", "", files_list)
filenames <- gsub(".txt", "", filenames)

#Dump cluster gene lists into a list element
geneclusters <- list()
for (i in 1:length(files)) {
  geneclusters[[i]] <- read.csv(files[i], header=TRUE, sep = "\t")
  colnames(geneclusters[[i]])[2] <- "Gene"
  names(geneclusters)[i] <- filenames[i]
  geneclusters[[i]] <- as.vector(geneclusters[[i]]$Gene)
}

#Filter above genes by those with uniprot accession number: 
library(readxl)
genemodels <- as.data.frame(read_xlsx("~/Documents/NIEHS_LSU_UCD/rnaseq/Ranalysis/Gene.Models.Conservative.and.Liberal.xlsx"))
names(genemodels) <- c("geneID", "name", "cons", "lib")
unip <- list()

for (i in 1:length(geneclusters)){
  unip[[i]] <- genemodels[genemodels$geneID %in% geneclusters[[i]],]$lib
  unip[[i]] <- unip[[i]][!is.na(unip[[i]])]
  names(unip)[i] <- names(geneclusters)[i]
}
#Assign same length to all vectors, then run cbind: 
n <- max(lengths(unip))
for (i in 1:length(unip)) {
  length(unip[[i]]) <- n
}
david <- do.call(cbind, unip)
write.table(david, file="~/Documents/NIEHS_LSU_UCD/rnaseq/david_results/180306_exp00_clustergenelists.txt", row.names = FALSE, col.names = TRUE, sep="\t", quote=FALSE)

```